src/main/kotlin/com/autonomousapps/internal/graph/GraphWriter.kt

```diff
@@ -38,4 +44,29 @@ internal object GraphWriter {
     }
     append("}")
   }
+
+  /**
+   * Returns the [graph] sorted into topological order, ascending. Each node in the graph is paired with its in-degree,
+   * or the number of dependents.
+   *
+   * TODO(tsr): not sure if including the in-degree has any value. What I really want is a set of batches that are
+   *  internally independent. Essentially, I want a work graph that is maximally parallelizable.
+   */
+  fun topological(graph: Graph<Coordinates>): String {
+    return Topological(graph, graph.root())
+      .order
+      .joinToString(separator = "\n") {
+        // ":foo:bar 0"
+        // ":other:baz 1"
+        it.maybeProjectCoordinates(buildPath).gav() + " ${graph.inDegree(it)}"
+      }
+  }
```

Comment:
@staktrace this is the main area of interest for us currently.

I can think of very expensive algorithms that check if there's a path from one node to another, and using that to create the batches, but I feel like there might also be some common(ish) graph algo that can also cut up a graph into mutually semi-independent chunks. I'll page through the lit, but if you can think of anything I'm all ears.

Reply:
Omg, I just realized that what I want is basically just the dominance tree. It already contains all the information necessary to go in order in a maximally concurrent way. At least, I think. Will test later!

Reply:
Yeah I think using the dominator tree of the dependency graph makes sense here.